### PR TITLE
Replaced utcnow and utcfromtimestamp functions

### LINKED ISSUE #13125: Replace utcnow and utcfromtimestamp function calls
`datetime.utcnow` and `utcfromtimestamp` will be deprecated in Python 3.12. We use both in our codebase and should replace them, see https://github.com/python/cpython/issues/103857.

### PR DESCRIPTION
**Pull request by rielAsh24**
The above functions have been replaced with datetime.now() and datetime.fromtimestamp() aligning to Python 3.12 deprecation

All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [x] issues: fixes #13125 
- [x] tests added / passed

### CODE REVIEW

**Comment by ianthomas23** [comment]
Thanks @rielAsh24, I will take a look at this on Monday unless someone else beats me to it.

File: .github/workflows/bokeh-ci.yml
Comment (ianthomas23) [inline .github/workflows/bokeh-ci.yml]: This isn't valid, it is expecting a single value here not a sequence; see https://github.com/bokeh/bokeh/actions/runs/5032752743. These `integration-tests` are currently disabled anyway (see line 176 above) so changes in this section won't have any effect.

Actually, please can you revert all changes to this file. We don't want to attempt testing on python `3.12-dev` in CI as most of our dependencies do not exist on Python 3.12 yet. Plus your code editor (or similar) has changed the formatting in this file, which aren't changes we want to see in an unrelated PR.

File: src/bokeh/server/views/ws.py
Comment (ianthomas23) [inline src/bokeh/server/views/ws.py]: We don't really want two separate imports of `datetime` here, just one should be sufficient. I'd would be inclined to keep the original import and change all new mentions of `timezone` to `dt.timezone`.

File: src/bokeh/server/views/ws.py
Comment (ianthomas23) [inline src/bokeh/server/views/ws.py]: According to https://docs.python.org/3/library/datetime.html#datetime.datetime.utctimetuple we can change all calls from `utctimetuple()` to just `timetuple()` now.

File: src/bokeh/util/serialization.py
Comment (ianthomas23) [inline src/bokeh/util/serialization.py]: Again, we only want a single `datetime` import.

**Review by ianthomas23 — CHANGES_REQUESTED** [review]
@rielAsh24 If you can revert the `bokeh-ci.yml` changes and address my other comments, the Bokeh CI should run and we will see if the existing tests pass.

I have a suspicion that we will be making a bunch of `datetimes` timezone-aware when before they were indifferent to the timezone and this may not be desirable. But let's get the CI back up and running first.

File: .github/workflows/bokeh-ci.yml
Comment (rielAsh24) [reply .github/workflows/bokeh-ci.yml]: For sure! I had noticed this when I was trying to setup my environment that those variable values had changed. Reverting this back to the previous version.

Regarding the formatting, I'm using vscode on mac. Do you suggest any settings I should change?

Comment (ianthomas23) [reply .github/workflows/bokeh-ci.yml]: I suggest you turn off any automatic formatting that you have enabled when you are making code changes. They are fine to have enabled for your own work, but when submitting changes to Open Source projects you don't want to change any other lines unnecessarily.

File: src/bokeh/server/views/ws.py
Comment (rielAsh24) [reply src/bokeh/server/views/ws.py]: @ianthomas23 Got you. Addressed these in the recent commit

File: src/bokeh/server/views/ws.py
Comment (rielAsh24) [reply src/bokeh/server/views/ws.py]: Added these too

File: src/bokeh/util/serialization.py
Comment (rielAsh24) [reply src/bokeh/util/serialization.py]: Modified the imports here too

**Comment by rielAsh24** [comment]
Hey @ianthomas23, I addressed the changes mentioned and was curious about the tests failing on GitHub. Is there something I need to work on?

**Comment by ianthomas23** [comment]
The test failures aren't your fault, a PR was merged to today that fixes these problems. Ideally you would update your local `branch-3.2` and rebase your `13125_replace_utcnow_and_utcfromtimestamp_functions` branch against `branch-3.2`. If that doesn't mean anything to you (I don't know how much you have used `git`) then I can explain the individual steps.

**Comment by rielAsh24** [comment]
> The test failures aren't your fault, a PR was merged to today that fixes these problems. Ideally you would update your local `branch-3.2` and rebase your `13125_replace_utcnow_and_utcfromtimestamp_functions` branch against `branch-3.2`. If that doesn't mean anything to you (I don't know how much you have used `git`) then I can explain the individual steps.

@ianthomas23 You're right. The rebase was required. Synced the changes to current `branch-3.2`. Thanks for that!

**Comment by bryevdv** [comment]
@rielAsh24 can you mention the command(s) that you issued? It looks like a merge rather than a rebase occurred, and that would be fine, but it seems like something has gone amiss since GH now reports 800 files changed in this PR.

**Comment by rielAsh24** [comment]
> @rielAsh24 can you mention the command(s) that you issued? It looks like a merge rather than a rebase occurred, and that would be fine, but it seems like something has gone amiss since GH now reports 800 files changed in this PR.

I did the rebase on Github Desktop:
It asked me to rebase against `branch-3.2`, and it displayed a conflict for file `src/bokeh/util/token.py`

Rebase_Github_Desktop

It displays merge onto the same branch as you mentioned. Later used git CLI to push to origin:
`git push -u origin 13125_replace_utcnow_and_utcfromtimestamp_functions`

Update: I checked the commits and it seemed to add all the previous commits to this pull requests. Do you suggest I change something?

Update2: @bryevdv I'm working on rebasing this back to branch and undo merge

**Comment by rielAsh24** [comment]
> > @rielAsh24 can you mention the command(s) that you issued? It looks like a merge rather than a rebase occurred, and that would be fine, but it seems like something has gone amiss since GH now reports 800 files changed in this PR.
> 
> I did the rebase on Github Desktop: It asked me to rebase against `branch-3.2`, and it displayed a conflict for file `src/bokeh/util/token.py`
> 
> Rebase_Github_Desktop
> 
> It displays merge onto the same branch as you mentioned. Later used git CLI to push to origin: `git push -u origin 13125_replace_utcnow_and_utcfromtimestamp_functions`
> 
> Update: I checked the commits and it seemed to add all the previous commits to this pull requests. Do you suggest I change something?
> 
> Update2: @bryevdv I'm working on rebasing this back to branch and undo merge

@bryevdv @ianthomas23 I have rebased `branch-13125_replace_utcnow_and_utcfromtimestamp_functions` to sync with `branch-3.2`.

**Comment by bryevdv** [comment]
Thanks @rielAsh24 the GH history looks more reasonable now. We are working on getting a 3.2 release out this week so just FYI it may be next week before we get back to looking at this for merge

**Review by ianthomas23 — APPROVED** [review]
Thanks @rielAsh24 . I am approving this PR but I'd like another maintainer to express an opinion too.

I have some concerns, not about the implementation but about the ramifications of the change which I think are larger than I originally considered. Previously we didn't deal with timezones, we left that up to the user. For example, if you had a `DatetimeRangeSlider` with a Python `on_change` callback you could call `slider.value_as_datetime` as you'd receive a `datetime` with no timezone. With this change the returned `datetime` has a timezone applied. This is technically an API change. Somebody will have purposefully or accidentally relied upon the old behaviour and receive a different result now.

We don't have the option of leaving the code as it was, so we either

1. Make this change and accept the addition of UTC as the timezone, or
2. Make this change but also explicitly drop the UTC timezone using `datetime.replace(tzinfo=None)`.

Option 2 is equivalent to what we do now whereas option 1 is where most of the Python community is going by dropping these two deprecated functions. I slightly prefer option 1 hence my approval here. But I am aware that I personally spend most of my life at or near UTC so I may be less aware of timezone problems than others.

**Comment by rielAsh24** [comment]
> Thanks @rielAsh24 . I am approving this PR but I'd like another maintainer to express an opinion too.
> 
> I have some concerns, not about the implementation but about the ramifications of the change which I think are larger than I originally considered. Previously we didn't deal with timezones, we left that up to the user. For example, if you had a `DatetimeRangeSlider` with a Python `on_change` callback you could call `slider.value_as_datetime` as you'd receive a `datetime` with no timezone. With this change the returned `datetime` has a timezone applied. This is technically an API change. Somebody will have purposefully or accidentally relied upon the old behaviour and receive a different result now.
> 
> We don't have the option of leaving the code as it was, so we either
> 
> 1. Make this change and accept the addition of UTC as the timezone, or
> 2. Make this change but also explicitly drop the UTC timezone using `datetime.replace(tzinfo=None)`.
> 
> Option 2 is equivalent to what we do now whereas option 1 is where most of the Python community is going by dropping these two deprecated functions. I slightly prefer option 1 hence my approval here. But I am aware that I personally spend most of my life at or near UTC so I may be less aware of timezone problems than others.

@ianthomas23 Thanks for approving! Glad I could contribute😄.

I completely understand the concern mentioned. Even while implementing these changes, I was trying to stick to `option 2`, as it was originally implemented in `branch-3.2` files. But the test results wanted me to implement `option 1`. Hope these changes work out well!

**Comment by rielAsh24** [comment]
Hey @ianthomas23 ! Request you to let me know if I need to tag/request a review from a maintainer to proceed further.

**Comment by ianthomas23** [comment]
This can wait until after the impending 3.2 release. I would normally avoid requesting a review from a specific person as that would be asking them to commit some of their time without asking them first, and could be considered bad etiquette.

**Comment by rielAsh24** [comment]
> This can wait until after the impending 3.2 release. I would normally avoid requesting a review from a specific person as that would be asking them to commit some of their time without asking them first, and could be considered bad etiquette.

Glad to know this! Looking forward to the release 😄

Thank you for guidance throughout, appreciate it.

**Comment by bryevdv** [comment]
@rielAsh24 @ianthomas23 Nominally I think this is probably OK, and given the deprecations we will have to do something at some point... If we must make a change, it's OK I think, as long as we adequately document the changes. I guess I am also just not completely certain about the scope / fallout though. Just to clarify, the main concern is with the change in `serialization.py` that adds a `tz` parameter, correct? 

Can we just make a couple of other comparisons between 3.2 and this branch? e.g. plots with datetime axes, datetime range sliders, etc. Just to make sure there aren't differences we have not accounted for?

**Comment by rielAsh24** [comment]
> @rielAsh24 @ianthomas23 Nominally I think this is probably OK, and given the deprecations we will have to do something at some point... If we must make a change, it's OK I think, as long as we adequately document the changes. I guess I am also just not completely certain about the scope / fallout though. Just to clarify, the main concern is with the change in `serialization.py` that adds a `tz` parameter, correct?
> 
> Can we just make a couple of other comparisons between 3.2 and this branch? e.g. plots with datetime axes, datetime range sliders, etc. Just to make sure there aren't differences we have not accounted for?

Sure! I will try comparing on Github for `branch-3.2` and `3.3`.

Regarding the range sliders and datetime axes, are these comparisons on the UI end?

**Comment by bryevdv** [comment]
> Regarding the range sliders and datetime axes, are these comparisons on the UI end?

For axes, I think just a visual comparison that they look the same on the same inputs (in a local timezone outside UTC) is sufficient. For the widgets, a nice extra step would be to look at one-line callbacks that print the slider values, etc, and make sure that the reported values are not getting changed in a way we don't expect.

**Comment by rielAsh24** [comment]
> > Regarding the range sliders and datetime axes, are these comparisons on the UI end?
> 
> For axes, I think just a visual comparison that they look the same on the same inputs (in a local timezone outside UTC) is sufficient. For the widgets, a nice extra step would be to look at one-line callbacks that print the slider values, etc, and make sure that the reported values are not getting changed in a way we don't expect.

Got you! Would surely look into this.

Also, while exploring the comparison, I discovered more files that have implemented the `datetime.now()` and `datetime.fromtimestamp()` functions, but have no `tz=datetime.timezone.utc` argument. Following are the files with the respective functions and the occurences:

### datetime.now()
first_steps_4_datetime_axis.py [1]
server_sent_events_source.py [1]
events.py [1]
test_aliases.py [1]
test_datetime.py [1]
test_sources.py [2]

### datetime.fromtimestamp()
test_daterange_slider.py [2]
test_dateslider.py [1]

Adding the `tz` argument invoked an error in only one file: `test_sources.py`, on the below lines:
https://github.com/bokeh/bokeh/blob/ee31f786164bc456efddd97e44686f8a74cb3a4b/tests/unit/bokeh/models/test_sources.py#L366

https://github.com/bokeh/bokeh/blob/ee31f786164bc456efddd97e44686f8a74cb3a4b/tests/unit/bokeh/models/test_sources.py#L383

Requesting your suggestions on the changes and if I should commit them to the PR.

**Comment by ianthomas23** [comment]
@rielAsh24 I've talked this through with @bryevdv and we are happy to merge it as it is, and I will update the documentation in a separate PR to not delay this any further.

Thanks for fixing this, and I sorry it has taken so long for us to get round to deciding about it.

**Comment by rielAsh24** [comment]
> @rielAsh24 I've talked this through with @bryevdv and we are happy to merge it as it is, and I will update the documentation in a separate PR to not delay this any further.
> 
> Thanks for fixing this, and I sorry it has taken so long for us to get round to deciding about it.

Hey @ianthomas23 thanks for merging this and I'm glad I could contribute!

**Pull request merged**
